src/riak_kv_wm_utils.erl

```diff
@@ -332,13 +335,74 @@ is_valid_referer(RD) ->
             false;
-        OriginTuple ->
-            true;
         RefererTuple ->
-            lager:debug("WM referrer not origin.  Origin ~p != Referer ~p\n", [OriginTuple, RefererTuple]),
-            false
+            allowable_origin(OriginTuple, RefererTuple)
+    end.
+
+%% @doc Register allowable origins (for CORS), see riak_kv.allowable_origin
+%% w/i riak_kv.schema.
+-spec register_allowable_origins([{atom(), [boolean()|{string(), integer()}]}]) -> ok.
+register_allowable_origins(Origins) ->
+    maybe_create_allowable_origins_ets(),
+    lager:info("Registering CORS Origins: ~p~n", [Origins]),
+    register_allowable_origins1(Origins).
+
+register_allowable_origins1([]) -> ok;
+register_allowable_origins1([{all, [true]}|_T]) ->
+    AllOrigins = normalize_referer('*'),
+    %% wrap the origin for full match on ets lookup
+    ets:insert(?ALLOWABLE_ORIGINS_TABLE, {AllOrigins});
+register_allowable_origins1([{all, _}|T]) ->
+    register_allowable_origins1(T);
+register_allowable_origins1([{http, HttpOrigins}|T]) ->
+    register_allowable_origins2(http, HttpOrigins),
+    register_allowable_origins1(T);
+register_allowable_origins1([{https, HttpOrigins}|T]) ->
+    register_allowable_origins2(https, HttpOrigins),
+    register_allowable_origins1(T).
+
+register_allowable_origins2(Scheme, HostPortPairs) ->
+    OriginsWild = [{normalize_referer({Scheme, Host, "*"})} || {Host} <- HostPortPairs],
+    Origins = [{normalize_referer({Scheme, Host, Port})} || {Host, Port} <- HostPortPairs],
+    ets:insert(?ALLOWABLE_ORIGINS_TABLE, OriginsWild ++ Origins).
+
+allowable_origins_table_exists() ->
+    undefined /= ets:info(?ALLOWABLE_ORIGINS_TABLE).
+
+maybe_create_allowable_origins_ets() ->
+    case allowable_origins_table_exists() of
+        false ->
+            ets:new(?ALLOWABLE_ORIGINS_TABLE,
+                    [named_table, public, ordered_set]);
+        _ ->
+            ok
     end.
 
+allowable_origin(OriginTuple, OriginTuple)  ->
+    true;
+allowable_origin(OriginTuple, RefererTuple) ->
+    RefererTuple1 = {Scheme, Host, Port} = normalize_referer(RefererTuple),
+    case allowable_origins_table_exists() andalso
+        (ets:member(?ALLOWABLE_ORIGINS_TABLE, RefererTuple1) orelse
+         ets:member(?ALLOWABLE_ORIGINS_TABLE, {Scheme, Host, "*"}) orelse
+         ets:member(?ALLOWABLE_ORIGINS_TABLE, {Scheme, "*", "*"}) orelse
+         ets:member(?ALLOWABLE_ORIGINS_TABLE, {Scheme, "*", Port}) orelse
+         ets:member(?ALLOWABLE_ORIGINS_TABLE, {"*", Host, Port}) orelse
+         ets:member(?ALLOWABLE_ORIGINS_TABLE, {"*", "*", "*"})) of
+        false ->
+            lager:debug("WM referrer not origin. Origin ~p != Referer ~p" ++
+                        " and Referer not in allowable origins.\n",
+                        [OriginTuple, RefererTuple]),
+            false;
+        _ -> true
+    end.
+
+normalize_referer('*') -> {"*", "*", "*"};
+normalize_referer({Scheme, "127.0.0.1", Port}) ->
```

Comment:
Pedants'R'Us.

Dredging up distant memories, this should really be 127.*. The entire `127` class A network is equivalent to localhost.